File: sample-oauth2-idp/src/ports/refresh-token-store.ts
/**
 * Layer 3 — Application Logic（ポート定義）
 *
 * リフレッシュトークンのストア。
 * ADR-004 のファミリー失効をサポートする。
 */

import type { RefreshTokenRecord } from '../spec-bindings/schemas'

export interface RefreshTokenStore {
  findByHash(hash: string): Promise<RefreshTokenRecord | null>
  save(record: RefreshTokenRecord): Promise<void>
  /**
   * トークンを「rotated」にマークし、新しいトークンレコードと不可分に保存する。
   * 並行ローテーションを防ぐため atomic 操作を期待する。
   */
  rotate(parentId: string, newRecord: RefreshTokenRecord): Promise<RefreshTokenRecord | null>
  /**
   * 指定 family_id のすべてのトークンを revoked にする。
   */
  revokeFamily(family_id: string): Promise<void>
}

File: ra-idp-go/internal/oauth2/ports/refresh_token_store.go
package ports

import (
	"context"

	"ra-idp-go/internal/spec"
)

type RefreshTokenStore interface {
	FindByHash(ctx context.Context, hash string) (*spec.RefreshTokenRecord, error)
	Save(ctx context.Context, rec *spec.RefreshTokenRecord) error
	// Rotate は parentId を rotated にしつつ新レコードを atomic に保存。
	Rotate(ctx context.Context, parentID string, newRec *spec.RefreshTokenRecord) (*spec.RefreshTokenRecord, error)
	RevokeFamily(ctx context.Context, familyID string) error
	// DeleteAllForSub は ADR-036 の anonymize cascade から呼ばれる。
	// 対象 sub の RefreshToken を物理削除する。
	DeleteAllForSub(ctx context.Context, sub string) error
}
